_posts/2023-04-30-writing-code-with-llm-dont-forget-kani.md

```diff
@@ -0,0 +1,306 @@
+
+We were also eager to explore the possibilities of using LLMs for daily coding tasks, so we decided to put ChatGPT to the test.
+In this blog post, we'll share our experiences with using ChatGPT and Kani to solve example coding problems, and explore how these technology could revolutionize the way we approach software development.
+The experiment we did include the following three tasks.
+
+1. We write prompt and generate (unverified!) code with ChatGPT.
+2. We verify the generated code to see if they are correct with Kani.
+3. We use both ChatGPT and Kani to generate verified code.
+
+Our result showed that, ChatGPT is easily to use and can efficient generate code from text description. 
+While the generated code is not always sound, program verifiers like Kani can effectively detect bugs in the generated code and help ChatGPT fix them.
+
+## Code Generation using ChatGPT
+We demonstrate the three tasks with two examples.
+Let's start our first task: can we write code using prompt with ChatGPT?
+
+#### Example 1.a
+The first example is a toy problem of writing a rust function to compute the average of two integers.
+So the prompt is also very simple and straightforward.
+
+```
+Q: Can you write me a rust function that compute the integer average of two integers?
+```
+
+Giving the above prompt, ChatGPT answered us with the following function
+```rust
+fn integer_average(a: i32, b: i32) -> i32 {
+    (a + b) / 2
+}
+```
+, the explanation of the function
+>In this function, we define a function called integer_average that takes two integer arguments a and b.
+The function then adds the two arguments together and divides the result by 2 to compute the integer average.
+
+, and even a drive function to call the average function.
+```rust
+fn main() {
+    let a = 4;
+    let b = 8;
+    let result = integer_average(a, b);
+    println!("The average of {} and {} is {}", a, b, result);
+}
+```
+
+#### Example 2.a
+The second example we tried is more like a real-world coding task.
+We wrote a rust function that compute the lowest unset bit in a `u32` integer.
+Our implementation uses a loop to iterate through all 32 bits in input.
+And we know that this could be slower than using some [bit twiddling hacks](https://graphics.stanford.edu/~seander/bithacks.html).
+So, we want to optimize the function with ChatGPT.
+The prompt we provided including the implementation we wrote.
+
+``` 
+Q: Optimize this rust function:
+fn lowest_unset_bit(x: u32) -> Option<u8> {
+    let mut mask = 1;
+    for i in 0..32 {
+        if x & mask == 0 {
+            return Some(i);
+        }
+        mask <<= 1;
+    }
+    None
+} 
+```
+
+After a few seconds, ChatGPT gave us a solution
+``` rust
+fn lowest_unset_bit(x: u32) -> Option<u8> {
+    if x == std::u32::MAX{
+        return None;
+    }
+    let y = x | (x + 1);
+    let z = y ^ (y + 1);
+    Some((z.trailing_zeros() % 32) as u8)
+}
+```
+with the explanation of it.
```

Comment:
```suggestion
along with an explanation.
```

Reply:
You don't include the explanation here? Might be appropriate since it's doing some bit magic.

Reply:
I took this example from @zhassan-aws 's video which doesn't show the full explanation.
@zhassan-aws Have you store the explanation of this answer? Or probably it can be found from the ChatGPT history.